onnx/defs/math/defs.cc

```diff
@@ -416,6 +471,117 @@ will throw errors.
   .TypeConstraint("T", { "tensor(float16)", "tensor(float)", "tensor(double)" },
         "Constrain input and output types to float tensors.");
 
+OPERATOR_SCHEMA(LogSoftmax)
+  .NumInputs(1)
+  .NumOutputs(1)
+  .SetDoc(R"DOC(
+The operator computes the log softmax normalized values for each layer in the batch
+ of the given input. The input is a 2-D tensor (Tensor<float>) of size
+(batch_size x input_feature_dimensions). The output tensor has the same shape
+and contains the softmax normalized values of the corresponding input.
+
+X does not need to explicitly be a 2D vector; rather, it will be
+coerced into one. For an arbitrary n-dimensional tensor
+X \in [a_0, a_1, ..., a_{k-1}, a_k, ..., a_{n-1}] and k is
+the axis provided, then X will be coerced into a 2-dimensional tensor with
+dimensions [a_0 * ... * a_{k-1}, a_k * ... * a_{n-1}]. For the default
+case where axis=1, this means the X tensor will be coerced into a 2D tensor
+of dimensions [a_0, a_1 * ... * a_{n-1}], where a_0 is often the batch size.
+In this situation, we must have a_0 = N and a_1 * ... * a_{n-1} = D.
+Each of these dimensions must be matched correctly, or else the operator
+will throw errors.
+)DOC")
+  .Attr("axis",
+        "(int) default to 1; describes the axis of the inputs when coerced "
+        "to 2D; defaults to one because the 0th axis most likely describes "
+        "the batch_size",
+        AttrType::INT)
+  .Input(0, "input",
+         "The input tensor that's coerced into a 2D matrix of size (NxD) "
+         "as described above.", "T")
+  .Output(0, "output", "The softmax normalized output values with the same "
+          "shape as input tensor.", "T")
+  .TypeConstraint("T", { "tensor(float16)", "tensor(float)", "tensor(double)" },
+        "Constrain input and output types to float tensors.");
+
+OPERATOR_SCHEMA(Hardmax)
+  .NumInputs(1)
+  .NumOutputs(1)
+  .SetDoc(R"DOC(
+The operator computes the hardmax values for each layer in the batch
+ of the given input. The input is a 2-D tensor (Tensor<float>) of size
+(batch_size x input_feature_dimensions). The output tensor has the same shape
+and contains the softmax normalized values of the corresponding input.
```

Comment:
so what exactly does this op do? I can see you say about softmax here, but I don't get it. The TF one with the same name has but does something different (effectively argmax)

Reply:
Hardmax([0.1, 0.2, 0.3]) == [0, 0, 1]. The largest entry will be mapped to 1 and other entries will be zero. 
This is not the same as argmax, where argmax only gives the index of largest entry.

Reply:
So it's basically argmax + onehot. Makes sense. Probably deserves a separate op